### PR TITLE
add info to customer table and stripe customer data

### PR DESCRIPTION
**Pull request by anamarn**
Solves (https://github.com/twentyhq/private-issues/issues/194)

**TLDR**

Updates the billingCustomer table data using stripe webhooks event, also updates the customer's metadata in stripe, in order to contain the workspaceId associated to this customer.

**In order to test**
Billing:

- Set IS_BILLING_ENABLED to true
- Add your BILLING_STRIPE_SECRET and BILLING_STRIPE_API_KEY
- Add your BILLING_STRIPE_BASE_PLAN_PRODUCT_ID (use the one in testMode > Base Plan)
- 

Authenticate with your account in the stripe CLI
Run the command: stripe listen --forward-to http://localhost:3000/billing/webhooks

Run the twenty workker

Authenticate yourself on the app choose a plan and run the app normally. In stripe and in posgress the customer table data should be added.

**Next steps**
Learn more about integrations tests and implement some for this PR.

### CODE REVIEW

File: packages/twenty-server/src/engine/core-modules/billing/billing.controller.ts [hunk 1]
```diff
@@ -61,16 +65,38 @@ export class BillingController {
         return;
       }
 
-      await this.billingWehbookService.processStripeEvent(
+      await this.billingWebhookSubscriptionService.processStripeSubscriptionEvent(
         workspaceId,
         event.data,
       );
+      const customer = await this.stripeService.getCustomer(
+        event.data.object.customer as string,
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/billing.controller.ts:73]: why do we need to cast as?

File: packages/twenty-server/src/engine/core-modules/billing/services/billing-webhook-customer.service.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+        stripeCustomerId: data.object.id,
+      },
+      {
+        conflictPaths: ['workspaceId', 'stripeCustomerId'],
+        skipUpdateIfNoValuesChanged: true,
+      },
+    );
+  }
+
+  async verifyStripeCustomerMetadata(stripeCustomer: Stripe.Customer) {
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/services/billing-webhook-customer.service.ts:46]: I think this function is not useful and does not do what we expect it to do (it's returning undefined sometimes, sometimes the workspaceId).

We could rename it to: getWorkspaceIdFromBillingCustomer

File: packages/twenty-server/src/engine/core-modules/billing/billing.controller.ts [hunk 3]
```diff
@@ -61,16 +65,38 @@ export class BillingController {
...
+          BillingExceptionCode.BILLING_CUSTOMER_DELETED,
+        );
+      }
+
+      const stripeMetadataNeedsUpdate =
+        await this.billingWebhookCustomerService.verifyStripeCustomerMetadata(
+          customer,
+        );
+
+      if (stripeMetadataNeedsUpdate) {
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/billing.controller.ts:88]: this is actually a workspaceId!

File: packages/twenty-server/src/engine/core-modules/billing/services/billing-webhook-customer.service.ts (hunk 2 above)
Comment (charlesBochet) [reply packages/twenty-server/src/engine/core-modules/billing/services/billing-webhook-customer.service.ts:46]: I feel we actually don't need this method, let's just move this logic to the subscription webhook service

File: packages/twenty-server/src/engine/core-modules/billing/billing.exception.ts [hunk 4]
```diff
@@ -11,4 +11,5 @@ export class BillingException extends CustomException {
 
 export enum BillingExceptionCode {
   BILLING_CUSTOMER_NOT_FOUND = 'BILLING_CUSTOMER_NOT_FOUND',
+  BILLING_CUSTOMER_DELETED = 'BILLING_CUSTOMER_DELETED',
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/billing.exception.ts:14]: why is it an issue, we don't want to update deleted customers subscription?

File: packages/twenty-server/src/engine/core-modules/billing/services/billing-webhook-customer.service.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+      | Stripe.CustomerDeletedEvent.Data
+      | Stripe.CustomerUpdatedEvent.Data,
+  ) {
+    const stripeMetadataWorkspaceId = data.object.metadata?.workspaceId;
+    const billingCustomer = await this.billingCustomerRepository.findOne({
+      where: { stripeCustomerId },
+    });
+
+    if (!billingCustomer && !stripeMetadataWorkspaceId) {
+      return; //do nothing
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/services/billing-webhook-customer.service.ts:29]: comment not useful!

Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/services/billing-webhook-customer.service.ts:29]: why do nothing btw, is this a normal behavior?

File: packages/twenty-server/src/engine/core-modules/billing/services/billing-webhook-customer.service.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    }
+    if (billingCustomer && !stripeMetadataWorkspaceId) {
+      return billingCustomer.workspaceId; //send stripe to update (stripe customer metadata is not yet up to date)
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/services/billing-webhook-customer.service.ts:32]: no comments!

File: packages/twenty-server/src/engine/core-modules/billing/services/billing-webhook-customer.service.ts [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+import { BillingCustomer } from 'src/engine/core-modules/billing/entities/billing-customer.entity';
+@Injectable()
+export class BillingWebhookCustomerService {
+  protected readonly logger = new Logger(BillingWebhookCustomerService.name);
+  constructor(
+    @InjectRepository(BillingCustomer, 'core')
+    private readonly billingCustomerRepository: Repository<BillingCustomer>,
+  ) {}
+
+  async processStripeCustomerEvent(
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/services/billing-webhook-customer.service.ts:16]: this function should not return a workspaceId (you need it because you are putting some other logic depending in the controller). If it does we should make the naming clear

File: packages/twenty-server/src/engine/core-modules/billing/services/billing-webhook-customer.service.ts (hunk 6 above)
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/services/billing-webhook-customer.service.ts:32]: no!

File: packages/twenty-server/src/engine/core-modules/billing/stripe/stripe.service.ts [hunk 8]
```diff
@@ -114,7 +114,10 @@ export class StripeService {
       success_url: successUrl,
       cancel_url: cancelUrl,
     });
-  }
+  } // I prefered to not create a customer with metadat before the checkout, because it would break the tax calculation
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/stripe/stripe.service.ts:117]: understand the comment but why is it in here?

File: packages/twenty-server/src/engine/core-modules/billing/utils/billing-webhook-entitlement-transformer.util.ts [hunk 9]
```diff
@@ -0,0 +1,22 @@
+import Stripe from 'stripe';
+
+import { BillingEntitlementKey } from 'src/engine/core-modules/billing/enums/billing-entitlement-key.enum';
+
+export const transformStripeEntitlementEventToEntitlementRepositoryData = (
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/utils/billing-webhook-entitlement-transformer.util.ts:5]: file name does not meet function name

Comment (charlesBochet) [reply packages/twenty-server/src/engine/core-modules/billing/utils/billing-webhook-entitlement-transformer.util.ts:5]: based on data type, looks the name should reflect the fact this is an UpdatedEvent

File: packages/twenty-server/src/engine/core-modules/billing/utils/billing-webhook-subscription-transformer.util.ts [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+export const transformStripeSubscriptionEventToSubscriptionRepositoryData = (
+  workspaceId: string,
+  data:
+    | Stripe.CustomerSubscriptionUpdatedEvent.Data
+    | Stripe.CustomerSubscriptionCreatedEvent.Data
+    | Stripe.CustomerSubscriptionDeletedEvent.Data,
+) => {
+  return {
+    workspaceId,
+    stripeCustomerId: data.object.customer as string,
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/utils/billing-webhook-subscription-transformer.util.ts:15]: as string?

File: packages/twenty-server/src/engine/core-modules/billing/utils/billing-webhook-subscription-transformer.util.ts [hunk 11]
```diff
@@ -0,0 +1,79 @@
...
+    cancelAt: data.object.cancel_at
+      ? new Date(data.object.cancel_at * 1000)
+      : undefined,
+    canceledAt: data.object.canceled_at
+      ? new Date(data.object.canceled_at * 1000)
+      : undefined,
+  };
+};
+
+export const transformStripeSubscriptionEventToSubscriptionItemRepositoryData =
```
Comment (charlesBochet) [inline packages/twenty-server/src/engine/core-modules/billing/utils/billing-webhook-subscription-transformer.util.ts:46]: one util, one file!

File: packages/twenty-server/src/engine/core-modules/billing/billing.controller.ts (hunk 1 above)
Comment (anamarn) [reply packages/twenty-server/src/engine/core-modules/billing/billing.controller.ts:73]: because the  _event.data.object.customer_ is a  _Stripe.Subscription.customer_ which can be a 
_string | Stripe.Customer | Stripe.DeletedCustomer_.
Thus we force the cast to string if we want to retrieve the customer from stripe using the stripeId (string)

File: packages/twenty-server/src/engine/core-modules/billing/stripe/stripe.service.ts (hunk 8 above)
Comment (anamarn) [reply packages/twenty-server/src/engine/core-modules/billing/stripe/stripe.service.ts:117]: because Felix suggested this method for adding the customer metadata

File: packages/twenty-server/src/engine/core-modules/billing/utils/billing-webhook-subscription-transformer.util.ts (hunk 10 above)
Comment (anamarn) [reply packages/twenty-server/src/engine/core-modules/billing/utils/billing-webhook-subscription-transformer.util.ts:15]: Because
data.object.customer is of type
(property) Stripe.Price.product: string | Stripe.Product | Stripe.DeletedProduct

This is the same method we use to upsert data in BillingSubscriptionItemRepository (it hasn't changed). Do we need to refactorize?

File: packages/twenty-server/src/engine/core-modules/billing/billing.controller.ts (hunk 1 above)
Comment (FelixMalfait) [reply packages/twenty-server/src/engine/core-modules/billing/billing.controller.ts:73]: `as` is a Typescript keyword, it doesn't really perform real casting as I think you're implying here. Agree with Charles that it feels like a code smell. Are you sure it isn't always returning a customer object and then you can access .id?

**Review by charlesBochet — APPROVED** [review]
very clear! bravo!
